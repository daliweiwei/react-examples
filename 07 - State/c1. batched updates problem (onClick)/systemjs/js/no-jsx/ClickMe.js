/*! European Union Public License version 1.2 !*/
/*! Copyright Â© 2018 Rick Beerendonk          !*/

import React from 'react';

class ClickMe extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: 0 };

    // Bind all non-react methods to this.
    this.handleClick = this.handleClick.bind(this);
  }
  handleClick() {
    // Updates are batched for performance reasons.
    // Therefore this won't work:
    this.setState({ count: this.state.count + 0.4 });
    this.setState({ count: this.state.count + 0.6 });
  }
  render() {
    return React.createElement(
      'button',
      { onClick: this.handleClick },
      'This button has been clicked ',
      this.state.count,
      ' times'
    );
  }
}

export default ClickMe;
