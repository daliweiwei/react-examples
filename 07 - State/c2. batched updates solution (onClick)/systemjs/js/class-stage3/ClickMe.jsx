/*! European Union Public License version 1.2 !*/
/*! Copyright Â© 2018 Rick Beerendonk          !*/

import React from 'react';

class ClickMe extends React.Component {
  // Proposal: https://github.com/tc39/proposal-class-fields
  // Support: http://kangax.github.io/compat-table/esnext/#test-instance_class_fields
  state = { count: 0 };
  handleClick = () => {
    // Updates are batched for performance reasons.
    // Therefore a callback should be used:
    this.setState(state => ({ count: state.count + 0.4 }));
    this.setState(state => ({ count: state.count + 0.6 }));
  };

  render() {
    return (
      <button onClick={this.handleClick}>
        This link has been clicked {this.state.count} times
      </button>
    );
  }
}

export default ClickMe;
