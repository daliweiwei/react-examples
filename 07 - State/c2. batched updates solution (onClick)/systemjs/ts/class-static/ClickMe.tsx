/*! European Union Public License version 1.2 !*/
/*! Copyright Â© 2018 Rick Beerendonk          !*/

import React from 'react';

interface IClickMeProps {}

interface IClickMeState {
  count: number;
}

class ClickMe extends React.Component<IClickMeProps, IClickMeState> {
  state: IClickMeState = { count: 0 };
  handleClick = () => {
    // Updates are batched for performance reasons.
    // Therefore a callback should be used:
    this.setState(state => ({ count: state.count + 0.4 }));
    this.setState(state => ({ count: state.count + 0.6 }));
  };
  render() {
    return (
      <button onClick={this.handleClick}>
        This link has been clicked {this.state.count} times
      </button>
    );
  }
}

export default ClickMe;
