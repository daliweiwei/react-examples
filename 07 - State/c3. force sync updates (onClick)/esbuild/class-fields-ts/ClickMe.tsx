/*! European Union Public License version 1.2 !*/
/*! Copyright Â© 2018 Rick Beerendonk          !*/

import { Component } from 'react';
import { flushSync } from 'react-dom';

interface ClickMeProps {}

interface ClickMeState {
  count: number;
}

class ClickMe extends Component<ClickMeProps, ClickMeState> {
  // Proposal: https://github.com/tc39/proposal-class-fields
  // Support: http://kangax.github.io/compat-table/esnext/#test-instance_class_fields
  state = { count: 0 };
  handleClick = () => {
    // Updates are batched for performance reasons.
    // Use ReactDOM.flushSync():
    flushSync(() => {
      this.setState({ count: this.state.count + 0.4 });
    });
    flushSync(() => {
      this.setState({ count: this.state.count + 0.6 });
    });
  };

  render() {
    return (
      <button onClick={this.handleClick}>
        This link has been clicked {this.state.count} times
      </button>
    );
  }
}

export default ClickMe;
