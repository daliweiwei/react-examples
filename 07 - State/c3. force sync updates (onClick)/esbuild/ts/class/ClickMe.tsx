/*! European Union Public License version 1.2 !*/
/*! Copyright Â© 2018 Rick Beerendonk          !*/

import { Component } from 'react';
import { flushSync } from 'react-dom';

interface ClickMeProps {}

interface ClickMeState {
  count: number;
}

class ClickMe extends Component<ClickMeProps, ClickMeState> {
  constructor(props) {
    super(props);
    this.state = { count: 0 };

    // Bind all non-react methods to this.
    this.handleClick = this.handleClick.bind(this);
  }
  handleClick() {
    // Updates are batched for performance reasons.
    // Use ReactDOM.flushSync():
    flushSync(() => {
      this.setState({ count: this.state.count + 0.4 });
    });
    flushSync(() => {
      this.setState({ count: this.state.count + 0.6 });
    });
  }
  render() {
    return (
      <button onClick={this.handleClick}>
        This link has been clicked {this.state.count} times
      </button>
    );
  }
}

export default ClickMe;
