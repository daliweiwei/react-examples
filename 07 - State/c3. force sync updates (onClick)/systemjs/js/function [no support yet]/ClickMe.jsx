/*! European Union Public License version 1.2 !*/
/*! Copyright Â© 2018 Rick Beerendonk          !*/

import React from 'react';
import ReactDOM from 'react-dom';

function ClickMe() {
  const [count, setCount] = React.useState(0);

  return (
    <button
      onClick={() => {
        ReactDOM.flushSync(() => setCount(count + 0.4));
        ReactDOM.flushSync(() => setCount(count + 0.6));
      }}
    >
      This button has been clicked {count} times
    </button>
  );
}

export default ClickMe;
