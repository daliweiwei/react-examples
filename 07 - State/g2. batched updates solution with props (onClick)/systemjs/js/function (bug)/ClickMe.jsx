/*! European Union Public License version 1.2 !*/
/*! Copyright Â© 2019 Rick Beerendonk          !*/

import React from 'react';

function ClickMe({ value, setValue }) {
  const latestValue = React.useRef();
  latestValue.current = value;

  const [count, setCount] = React.useState(0);

  function handleClick() {
    // Updates are batched for performance reasons.
    // PROBLEM:
    // - In class components, the Parent class will render before setState callback is called
    // - In function components, "setState" callback is called before render.
    //   That invalidates "latestValue = useRef" as a solution.

    // Set property
    setValue(0.4);

    // Use property to set state
    setCount(c => {
      //debugger;
      return c + 0.6 + latestValue.current;
    });
  }

  return (
    <button onClick={handleClick}>
      This button has been clicked {count} times
    </button>
  );
}

export default ClickMe;
