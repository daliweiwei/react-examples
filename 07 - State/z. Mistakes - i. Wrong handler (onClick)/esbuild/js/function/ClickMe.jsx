/*! European Union Public License version 1.2 !*/
/*! Copyright Â© 2019 Rick Beerendonk          !*/

import React from 'react';

function ClickMe() {
  const [count, setCount] = React.useState(0);

  return (
    // setCount is not called when the event takes place.
    // The result of setCount() is called when the event takes place.
    <button onClick={/* () => */ setCount(count + 1)}>
      This button has been clicked {count} times
    </button>
  );
}

export default ClickMe;
