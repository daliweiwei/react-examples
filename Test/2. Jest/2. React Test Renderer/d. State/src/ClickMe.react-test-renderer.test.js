import React from 'react';
import renderer from 'react-test-renderer';
import ClickMe from './ClickMe';

/// React Test Renderer ///

describe('ClickMe', () => {
  it('renders 0 snapshot (react-test-renderer)', () => {
    const component = renderer.create(<ClickMe />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders 1 after click snapshot (react-test-renderer)', () => {
    const component = renderer.create(<ClickMe />);
    let tree = component.toJSON();
    tree.props.onClick();
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
