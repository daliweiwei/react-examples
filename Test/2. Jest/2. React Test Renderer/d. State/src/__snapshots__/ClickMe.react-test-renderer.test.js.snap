// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickMe renders 0 snapshot (react-test-renderer) 1`] = `
<a
  className="clickme"
  href="#"
  onClick={[Function]}
>
  This link has been clicked 
  0
   times
</a>
`;

exports[`ClickMe renders 1 after click snapshot (react-test-renderer) 1`] = `
<a
  className="clickme"
  href="#"
  onClick={[Function]}
>
  This link has been clicked 
  1
   times
</a>
`;
